### ISSUE TITLE
[Suggestion] Craft length limits and direction-based speed

### ISSUE THREAD

**Issue by Firecore-** [msummers123/Movecraft-3#51]
A maximum limit we can choose for how long/wide/tall a craft can be depending which way the pilot sign is facing.
Possibly a minimal limit too.

If you are able to detect sides, depending on which way the pilot sign is facing, perhaps moving to the side and/or rear can be slower than moving forward.

**Comment by Pocketkid2** [comment]
Based on direction of a cruise sign?

**Comment by TylerS1066** [comment]
I do like the idea of cruise sign direction, that would also allow us to implement a forward only cruise sign

**Comment by mwkaicz** [comment]
Just one question: how many ships you have where the is wide bigger than length?
Why Just don't add optional detection what's bigger?

**Comment by mwkaicz** [comment]
Detection must have some bigger diference offset. Or interresting could be also something like counting simple aerodynamic

**Comment by msummers123** [comment]
One of the more famous ships that is wider than it is long is this one:
http://www.planetminecraft.com/project/working-air-warship-silver-empire-assault-craft-with-adjustable-arc-cannons-and-aa-guns/

With its wide body it has more lift, and is able to carry heavy frontal armor. Also, in my humble opinion, it looks cool.

Anyway, the point is that if I base ship direction sensing off of geometry, I will get dozens of complaints about it "not working properly". Even if I went with the "longer than it is wide" assumption, it still doesn't give us which way is front.

Probably the best idea is to base it off the cruise sign like Tyler said.

Although, I have some other ideas I am working on. On our survival server laggy ships have become a problem. I thought I could limit it based on size and be fine. But the fact is that some 4k ships are much more laggy than a 10k ship.

Now the theory is to make it so that if moving the ship is taking a long time, it automatically slows down its movement. So instead of the whole server lagging, the ship slows down.

That actually is kind of based on aerodynamics. Because the movement engine can handle long aerodynamic shapes better than spread out non-aerodynamic ones, so they will naturally go faster. For a similar reason, ships will automatically go faster straight forward than side to side and particularly turning. 

It's an interesting idea and I hope I can get it to work.

**Comment by mwkaicz** [comment]
Problem with differences between 10k and 4k ship comes from structure of the ship and the direction where it's going. Because is diffrence if you are overwriting a block with other material or with the same. Also holes in the ship are taken as other material. Best speed has simetric cube.

But back to the topic: As @msummers123 said, it's too much easy to remove pilot sing and put it in another direction. We must use some system which will not be easy to rebuild.

About dynamic speed: It's good idea, but how you can detect, that slowing down causes that concrete ship? Yes, you can make some ratio between blocks count and the time between start and the end of translation task. But you must also count map updating process and nothing can tells you if the source of delay is the ship or some other thread/service running on the server. ... just my opinion.


Another option:
If we will be able to detect sails direction (some engine block isn't problem for ships without sails), we can take the ship direction from sails or engine. As I know, there are only two types of sails and both of them can be detected just only from the mast and yard structure  **T** (from front view)   and **L** (from side view)  ... i know **L** should be also from front view but it's not accustomed to build sails in minecraft by this way)  ... and the sail is necessary in **T** just only for detection which side is front side.

Then the craft file can contains this:
*needSails: false/true*
*sailsMaterial: WOOL*
*engineMaterial: REDSTONE_BLOCK*
*frontEngine: true/false*   (false means rear engine only when engineMaterial is set)


One question for all: should have a ship same speed to backward as side-speed?

**Comment by Pocketkid2** [comment]
What about repair signs? Nobody wants those to change direction. Oh by the way I heard that you can use those to clone ships (in a way)

**Comment by mwkaicz** [comment]
It must be something necessary, not optional. This causes dependency on WE.

**Comment by msummers123** [comment]
Yes, it would be more accurate to say you can use repair signs to mod a ship, because the repairs are limited to 25% of the ship's size. But you can, for example, manufacture small ships using a big ship. To me that's reasonable. You still have to have the component parts, and you still have to pay a per-block fee.
